x-pack/plugins/security/server/saved_objects/saved_objects_security_extension.ts

```diff
@@ -139,33 +507,109 @@ export class SavedObjectsSecurityExtension implements ISavedObjectsSecurityExten
     return { typeMap, status: 'unauthorized' };
   }
 
-  enforceAuthorization<A extends string>(params: EnforceAuthorizationParams<A>): void {
-    const { typesAndSpaces, action, typeMap, auditCallback } = params;
+  private auditHelper(params: AuditHelperParams | UpdateSpacesAuditHelperParams) {
+    const { action, useSuccessOutcome, objects, error, addToSpaces, deleteFromSpaces } =
+      params as UpdateSpacesAuditHelperParams;
+    // If there are no objects, we at least want to add a single audit log for the action
+    const toAudit = !!objects && objects?.length > 0 ? objects : ([undefined] as undefined[]);
+    for (const obj of toAudit) {
+      this.addAuditEvent({
+        action,
+        ...(!!obj && { savedObject: { type: obj.type, id: obj.id } }),
+        error,
+        // By default, if authorization was a success the outcome is 'unknown' because the operation has not occurred yet
+        // The GET action is one of the few exceptions to this, and hence it passes true to useSuccessOutcome
+        ...(!error && { outcome: useSuccessOutcome ? 'success' : 'unknown' }),
+        addToSpaces,
+        deleteFromSpaces,
+      });
+    }
+  }
+
+  /**
+   * The enforce method uses the result of an authoriation check authorization map) and a map
+   * of types to spaces (type map) to determinthe if the action is authorized for all types and spaces
+   * within the type map. If unauthorized for any type this method will throw.
+   */
+  private enforceAuthorization<A extends string>(params: EnforceAuthorizationParams<A>): void {
+    const { typesAndSpaces, action, typeMap, auditOptions } = params;
+    const {
+      objects: auditObjects,
+      bypass = 'never', // default for bypass
+      useSuccessOutcome,
+      addToSpaces,
+      deleteFromSpaces,
+    } = (auditOptions as UpdateSpacesAuditOptions) ?? {};
```

Comment:
`UpdateSpacesAuditOptions` extends `auditOptions`, and what I was hoping to achieve here was a way of passing in some number of interfaces that extend the base `auditOptions` interface.

If the parameter is not a `UpdateSpacesAuditOptions`, will `addToSpaces` and `deleteFromSpaces` not just be undefined? Is there a better coding pattern for achieving this? Am I over complicating it?

Reply:
I see that `auditOptions` has the type that's a type union of `AuditOptions` and `UpdateSpacesAuditOptions`, but `UpdateSpacesAuditOptions` extends `AuditOptions` and adds only _optional_ fields meaning that this union is technically almost equal to just `UpdateSpacesAuditOptions`. So if I'm not missing anything we can just use `auditOptions: UpdateSpacesAuditOptions` and we won't need to force-cast anything.

If we go this way, then I'd probably change a type name a bit, from `UpdateSpacesAuditOptions` (== "audit options for the update spaces action\something" or it just me who reads it like that?) to something like `AuthorizationAuditOptions` ("extended audit options for authz related operations") or `????AuditOptions` (== "audit options that might contain additional fields to space list manipulations").

What do you think?

Reply:
I just flattened to one interface and added some additional comments. I think I was overcomplicating it, especially since this is used only internal to the extension.